go/p2p/peermgmt/backup.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/common/persistent"
+)
+
+const (
+	// storeBucketName is the name of the bucket in which backup data is stored.
+	storeBucketName = "p2p/peer_manager"
+
+	// storePeerstoreKey is the bucket key under which peers from the peerstore are stored.
+	storePeerstoreKey = "peerstore"
+
+	// backupStartupDelay is the startup time delay for backups.
+	backupStartupDelay = 15 * time.Minute
+
+	// backupInterval is the time interval between backups.
+	backupInterval = 15 * time.Minute
+
+	// maxRestorePeers is the maximum number of peers connected on startup from the backup.
+	maxRestorePeers = 100
+)
+
+type peerstoreBackup struct {
+	logger *logging.Logger
+
+	host host.Host
+
+	store   *persistent.CommonStore
+	storeCh chan struct{}
+}
+
+func newPeerstoreBackup(h host.Host, cs *persistent.CommonStore) *peerstoreBackup {
+	l := logging.GetLogger("p2p/peer-manager/backup")
+
+	return &peerstoreBackup{
+		logger:  l,
+		host:    h,
+		store:   cs,
+		storeCh: make(chan struct{}, 1),
+	}
+}
+
+// Store persists peers to the store.
+func (b *peerstoreBackup) Store() error {
+	if b.store == nil {
+		return nil
+	}
+
+	// Allow only one store/load at a time.
+	b.storeCh <- struct{}{}
+	defer func() { <-b.storeCh }()
+
+	// Save only connected peers.
+	peers := b.host.Network().Peers()
+
+	if len(peers) == 0 {
+		return nil
+	}
+
+	b.logger.Debug("storing peers",
+		"count", len(peers),
+	)
+
+	infos := make([][]byte, 0, len(peers))
+	for _, p := range peers {
+		info := b.host.Peerstore().PeerInfo(p)
+		if len(info.Addrs) == 0 {
+			continue
+		}
+
+		json, err := info.MarshalJSON()
+		if err != nil {
+			return err
+		}
+
+		infos = append(infos, json)
+	}
+
+	// Persist addresses.
+	bucket, err := b.store.GetServiceStore(storeBucketName)
+	if err != nil {
+		return err
+	}
+	if err = bucket.PutCBOR([]byte(storePeerstoreKey), infos); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Load loads peers from the store.
+func (b *peerstoreBackup) Load() ([]*peer.AddrInfo, error) {
+	if b.store == nil {
+		return []*peer.AddrInfo{}, nil
+	}
+
+	// Allow only one store/load at a time.
+	b.storeCh <- struct{}{}
+	defer func() { <-b.storeCh }()
+
+	b.logger.Debug("loading peers")
+
+	// Load addresses in json form.
+	bucket, err := b.store.GetServiceStore(storeBucketName)
+	if err != nil {
+		return nil, err
+	}
+	var jsons [][]byte
+	if err = bucket.GetCBOR([]byte(storePeerstoreKey), &jsons); err != nil {
+		return nil, err
+	}
+
+	// Convert them.
+	infos := make([]*peer.AddrInfo, len(jsons))
+	for i, json := range jsons {
+		var info peer.AddrInfo
+		if err = info.UnmarshalJSON(json); err != nil {
+			return nil, err
+		}
+		infos[i] = &info
+	}
+
+	return infos, nil
+}
+
+// Backup periodically backups peers to the store.
+func (b *peerstoreBackup) Backup(ctx context.Context) {
```

Comment:
> I would suggest renaming this to somehing like `StartBackups` (or just `Start`).

Agree, I chose `Run` as `Start` would expect `Stop`and to avoid stuttering. 

> a guard against spawning multiple backup routines

Agree, added a waiting line.

> start a backup worker in a separate gorutine

Yes, could start a backup worker but in this case I decided that all peer manager's workers, handlers, watchers, ... use a blocking `Run(ctx)` instead of a non-blocking pair `Start`/`Stop`. Not sure what is better.

> backup worker would automatically be started in `newPeerstoreBackup`

I would never do that (unless I'm forced) for multiple reasons:
- Constructor should set initial values if non-zero values don't suffice and nothing more.
- I want to have control of all go-routines that are started so that I can do a graceful shutdown.
- I don't want to hide go routines. If I spawn them everywhere it will be hard to figure out how to stop them.
- Harder to write tests.
- It's ugly, see `libp2p` library how `host` starts immediately without our control.